### PR TITLE
Add domain-specific codeowners

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
Add domain-specific codeowners for the repository.
Added the entire eng team as owners for the entire repo (except for critical portions) to allow anyone on Eng to approve most PRs. Also added specific codeowners for specific portions of the repository, based on domain knowledge.

### CODE REVIEW

**Review by lupesko — APPROVED** [review]
LGTM. Thank you for putting this in!

File: CODEOWNERS [hunk 1]
```diff
@@ -1,8 +1,41 @@
+# By default, require admin approval for files that do not match any other rules
+* @mosaicml/composer-team-admins
+
 # Require admin approval to modify all files in the root of the repository
+# This includes setup.py, the README, and the CODEOWNERS file itself!
 /* @mosaicml/composer-team-admins
 
 # Require admin approval to change the CI build configuration
+# All CI Changes should be reviewed for security
```
Comment (bandish-shah) [inline CODEOWNERS:9]: Let's add me explicitly here as well for now.

File: CODEOWNERS [hunk 2]
```diff
@@ -1,8 +1,41 @@
...
 /.ci/ @mosaicml/composer-team-admins
 
 # Require admin approval to change the Dockerfile configuration
-/docker/ @mosaicml/composer-team-admins
+# All dockerfile changes should be reviewed for security
+# Adding Bandish explicitely, too.
```
Comment (moinnadeem) [inline CODEOWNERS:14]: sorry bud, these typos were hurting me

```suggestion
# Adding Bandish explicitly, too.
```

File: CODEOWNERS [hunk 3]
```diff
@@ -1,8 +1,41 @@
...
+/docs @mosaicml/composer-team-eng
+/examples @mosaicml/composer-team-eng
+/scripts @mosaicml/composer-team-eng
+/tests @mosaicml/composer-team-eng
+
+# Assigning domain-specific owners for certain portions of the repo so they will be
+# explicitely added in pull requests. Including @mosaicml/composer-team-eng
+# as an owner for all sections, so anyone on Composer Eng can approve any Composer PR
+# According to the CODEOWNER docs, the last match takes precedence, so @mosaicml/composer-team-eng
+# must be mentioend for each rule below.
```
Comment (moinnadeem) [inline CODEOWNERS:29]:
```suggestion
# must be mentioned for each rule below.
```

File: CODEOWNERS [hunk 4]
```diff
@@ -1,8 +1,41 @@
...
+/composer/algorithms/ @mosaicml/composer-team-research @dskhudia @mosaicml/composer-team-eng
+/composer/cli/ @jbloxham @mvpatel2000 @mosaicml/composer-team-eng
+/composer/datasets/ @mosaicml/composer-team-research @dskhudia @mosaicml/composer-team-eng
```
Comment (bandish-shah) [inline CODEOWNERS:32]: Should James just own everything under `/composer/datasets/`?

**Review by bandish-shah — APPROVED** [review]
Overall LGTM, few comments inline for discussion.

Additionally:
1. Daya is listed explicitly for majority of the Composer components, should this be the case?  Probably fine for now, we can always diversify going forward.
2. Having the research team listed is definitely good but I'm wondering if this is going to create too much noise?  Should we eventually find explicit PoC's on the research team we can tag vs blasting the whole team?

File: CODEOWNERS [hunk 5]
```diff
@@ -1,8 +1,41 @@
+# By default, require admin approval for files that do not match any other rules
+* @mosaicml/composer-team-admins
+
 # Require admin approval to modify all files in the root of the repository
+# This includes setup.py, the README, and the CODEOWNERS file itself!
 /* @mosaicml/composer-team-admins
 
 # Require admin approval to change the CI build configuration
+# All CI Changes should be reviewed for security
 /.ci/ @mosaicml/composer-team-admins
```
Comment (ravi-mosaicml) [inline CODEOWNERS:10]:
```suggestion
/.ci/ @bandish-shah @mosaicml/composer-team-admins
```

File: CODEOWNERS (hunk 4 above)
Comment (ravi-mosaicml) [reply CODEOWNERS:32]: A lot of the datasets are research specific, so I was thinking just streaming (which will eventually move to another repo).

File: CODEOWNERS (hunk 1 above)
Comment (ravi-mosaicml) [reply CODEOWNERS:9]: Done!

**Comment by lupesko** [comment]
> 1. Daya is listed explicitly for majority of the Composer components, should this be the case?  Probably fine for now, we can always diversify going forward.
> 2. Having the research team listed is definitely good but I'm wondering if this is going to create too much noise?  Should we eventually find explicit PoC's on the research team we can tag vs blasting the whole team?

These are good points. How about we merge this PR now, and figure out both #1 and #2 as a follow up? @bandish-shah you can take the lead on figuring both out.

File: CODEOWNERS (hunk 4 above)
Comment (ravi-mosaicml) [reply CODEOWNERS]: Going to leave as-is per discussion

**Pull request merged**